In [ ]:
!wget https://d2beiqkhq929f0.cloudfront.net/public_assets/assets/000/101/214/original/amazon_data4.zip


--2026-05-29 15:43:13--  https://d2beiqkhq929f0.cloudfront.net/public_assets/assets/000/101/214/original/amazon_data4.zip
Resolving d2beiqkhq929f0.cloudfront.net (d2beiqkhq929f0.cloudfront.net)... 3.175.29.13, 3.175.29.87, 3.175.29.125, ...
Connecting to d2beiqkhq929f0.cloudfront.net (d2beiqkhq929f0.cloudfront.net)|3.175.29.13|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1943535 (1.9M) [application/zip]
Saving to: ‘amazon_data4.zip’

amazon_data4.zip    100%[===================>]   1.85M  --.-KB/s    in 0.07s   

2026-05-29 15:43:14 (28.0 MB/s) - ‘amazon_data4.zip’ saved [1943535/1943535]



In [ ]:
!unzip amazon_data4.zip

Archive:  amazon_data4.zip
  inflating: products.csv            
  inflating: orders.csv              


In [ ]:
import pandas as pd

In [ ]:
orders= pd.read_csv('/content/orders.csv')

In [ ]:
products= pd.read_csv('/content/products.csv')

In [ ]:
orders.head()

,product_id,user_id,user_name,review_id,review_title,review_content,order_timestamp
0,B07JW9H4J1,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,2023-01-01 00:00:00.000000000
1,B098NS6PVG,"AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...","ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...","RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...","A Good Braided Cable for Your Type C Device,Go...",I ordered this cable to connect my phone to An...,2023-01-01 05:58:01.967213114
2,B096MSW6CT,"AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...","Kunal,Himanshu,viswanath,sai niharka,saqib mal...","R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...","Good speed for earlier versions,Good Product,W...","Not quite durable and sturdy,https://m.media-a...",2023-01-01 11:56:03.934426229
3,B08HDJ86NZ,"AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...","Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...","R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...","Good product,Good one,Nice,Really nice product...","Good product,long wire,Charges good,Nice,I bou...",2023-01-01 17:54:05.901639344
4,B08CF3B7N1,"AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...","rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...","R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...","As good as original,Decent,Good one for second...","Bought this instead of original apple, does th...",2023-01-01 23:52:07.868852459


In [ ]:
products.head()

,product_id,product_name,category,discounted_price,actual_price,rating,rating_count,about_product,img_link,product_link
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories,₹399,"₹1,099",4.2,24269.0,High Compatibility : Compatible With iPhone 12...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories,₹199,₹349,4.0,43994.0,"Compatible with all Type C enabled devices, be...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Ambrane-Unbreakable-Char...
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories,₹199,"₹1,899",3.9,7928.0,【 Fast Charger& Data Sync】-With built-in safet...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Sounce-iPhone-Charging-C...
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories,₹329,₹699,4.2,94363.0,The boAt Deuce USB 300 2 in 1 cable is compati...,https://m.media-amazon.com/images/I/41V5FtEWPk...,https://www.amazon.in/Deuce-300-Resistant-Tang...
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories,₹154,₹399,4.2,16905.0,[CHARGE & SYNC FUNCTION]- This cable comes wit...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Portronics-Konnect-POR-1...


### Data Overview: `orders` DataFrame

In [ ]:
print("Orders DataFrame Info:")
orders.info()
print("\nOrders DataFrame Description:")
orders.describe()

Orders DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   product_id       1465 non-null   object
 1   user_id          1465 non-null   object
 2   user_name        1465 non-null   object
 3   review_id        1465 non-null   object
 4   review_title     1465 non-null   object
 5   review_content   1465 non-null   object
 6   order_timestamp  1465 non-null   object
dtypes: object(7)
memory usage: 80.2+ KB

Orders DataFrame Description:


,product_id,user_id,user_name,review_id,review_title,review_content,order_timestamp
count,1465,1465,1465,1465,1465,1465,1465
unique,1351,1194,1194,1194,1194,1212,1465
top,B083342NKJ,"AHIKJUDTVJ4T6DV6IUGFYZ5LXMPA,AE55KTFVNXYFD5FPY...","$@|\|TO$|-|,Sethu madhav,Akash Thakur,Burger P...","R3F4T5TRYPTMIG,R3DQIEC603E7AY,R1O4Z15FD40PV5,R...","Worked on iPhone 7 and didn’t work on XR,Good ...","I am not big on camera usage, personally. I wa...",2023-12-31 00:00:00.000000000
freq,3,10,10,10,10,8,1


### Data Overview: `products` DataFrame

In [ ]:
print("Products DataFrame Info:")
products.info()
print("\nProducts DataFrame Description:")
display(products.describe(include='all'))

Products DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   product_id        1465 non-null   object 
 1   product_name      1465 non-null   object 
 2   category          1455 non-null   object 
 3   discounted_price  1453 non-null   object 
 4   actual_price      1457 non-null   object 
 5   rating            1449 non-null   float64
 6   rating_count      1463 non-null   float64
 7   about_product     1465 non-null   object 
 8   img_link          1465 non-null   object 
 9   product_link      1465 non-null   object 
dtypes: float64(2), object(8)
memory usage: 114.6+ KB

Products DataFrame Description:


,product_id,product_name,category,discounted_price,actual_price,rating,rating_count,about_product,img_link,product_link
count,1465,1465,1455,1453,1457,1449.000000,1463.000000,1465,1465,1465
unique,1351,1337,9,547,448,NaN,NaN,1293,1412,1465
top,B083342NKJ,"Fire-Boltt Ninja Call Pro Plus 1.83"" Smart Wat...",Electronics,₹199,₹999,NaN,NaN,[CHARGE & SYNC FUNCTION]- This cable comes wit...,https://m.media-amazon.com/images/I/413sCRKobN...,https://www.amazon.in/Borosil-Jumbo-1000-Watt-...
freq,3,5,524,53,120,NaN,NaN,6,3,1
mean,NaN,NaN,NaN,NaN,NaN,4.096618,18295.541353,NaN,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,0.291161,42753.864952,NaN,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,2.000000,2.000000,NaN,NaN,NaN
25%,NaN,NaN,NaN,NaN,NaN,4.000000,1186.000000,NaN,NaN,NaN
50%,NaN,NaN,NaN,NaN,NaN,4.100000,5179.000000,NaN,NaN,NaN
75%,NaN,NaN,NaN,NaN,NaN,4.300000,17336.500000,NaN,NaN,NaN


In [ ]:

samples_orders = pd.DataFrame({
    'product_id': ['P001', 'P002', 'P003'],
    'user_id': ['U001', 'U002', 'U003'],
    'review_id': ['R001', 'R002', 'R003']
})

samples_products = pd.DataFrame({
    'product_id': ['P001', 'P002', 'P004'],
    'product_name': ['Product A', 'Product B', 'Product D'],
    'category': ['Category 1', 'Category 2', 'Category 3']
})

In [ ]:
samples_orders

,product_id,user_id,review_id
0,P001,U001,R001
1,P002,U002,R002
2,P003,U003,R003


In [ ]:
samples_products

,product_id,product_name,category
0,P001,Product A,Category 1
1,P002,Product B,Category 2
2,P004,Product D,Category 3


In [ ]:
#pandas Merge

pd.merge(samples_orders, samples_products, on='product_id', how='inner')

,product_id,user_id,review_id,product_name,category
0,P001,U001,R001,Product A,Category 1
1,P002,U002,R002,Product B,Category 2


In [ ]:
pd.merge(samples_orders, samples_products, on='product_id', how='left')

,product_id,user_id,review_id,product_name,category
0,P001,U001,R001,Product A,Category 1
1,P002,U002,R002,Product B,Category 2
2,P003,U003,R003,NaN,NaN


In [ ]:
pd.merge(samples_orders, samples_products, on='product_id', how='right')

,product_id,user_id,review_id,product_name,category
0,P001,U001,R001,Product A,Category 1
1,P002,U002,R002,Product B,Category 2
2,P004,NaN,NaN,Product D,Category 3


In [ ]:
pd.merge(samples_orders, samples_products, on='product_id', how='outer') #best of both worlds

,product_id,user_id,review_id,product_name,category
0,P001,U001,R001,Product A,Category 1
1,P002,U002,R002,Product B,Category 2
2,P003,U003,R003,NaN,NaN
3,P004,NaN,NaN,Product D,Category 3


In [ ]:
df= pd.merge(orders, products, on='product_id', how='left')

In [ ]:
df

,product_id,user_id,user_name,review_id,review_title,review_content,order_timestamp,product_name,category,discounted_price,actual_price,rating,rating_count,about_product,img_link,product_link
0,B07JW9H4J1,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,2023-01-01 00:00:00.000000000,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories,₹399,"₹1,099",4.2,24269.0,High Compatibility : Compatible With iPhone 12...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...
1,B07JW9H4J1,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,2023-01-01 00:00:00.000000000,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories,₹399,"₹1,099",4.2,24270.0,High Compatibility : Compatible With iPhone 12...,https://m.media-amazon.com/images/I/51UsScvHQN...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...
2,B07JW9H4J1,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,2023-01-01 00:00:00.000000000,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories,₹399,"₹1,099",4.2,24269.0,High Compatibility : Compatible With iPhone 12...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...
3,B098NS6PVG,"AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...","ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...","RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...","A Good Braided Cable for Your Type C Device,Go...",I ordered this cable to connect my phone to An...,2023-01-01 05:58:01.967213114,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories,₹199,₹349,4.0,43994.0,"Compatible with all Type C enabled devices, be...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Ambrane-Unbreakable-Char...
4,B098NS6PVG,"AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...","ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...","RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...","A Good Braided Cable for Your Type C Device,Go...",I ordered this cable to connect my phone to An...,2023-01-01 05:58:01.967213114,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories,₹199,₹349,4.0,43993.0,"Compatible with all Type C enabled devices, be...",https://m.media-amazon.com/images/I/31zOsqQOAO...,https://www.amazon.in/Ambrane-Unbreakable-Char...
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1732,B08L7J3T31,"AHITFY6AHALOFOHOZEOC6XBP4FEA,AFRABBODZJZQB6Z4U...","Prabha ds,Raghuram bk,Real Deal,Amazon Custome...","R3G3XFHPBFF0E8,R3C0BZCD32EIGW,R2EBVBCN9QPD9R,R...","Received the product without spanner,Excellent...","I received product without spanner,Excellent p...",2023-12-30 00:07:52.131147540,Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...,Home&Kitchen,₹379,₹919,4.0,1090.0,SUPREME QUALITY 90 GRAM 3 LAYER THIK PP SPUN F...,https://m.media-amazon.com/images/I/41fDdRtjfx...,https://www.amazon.in/Noir-Aqua-Spanner-Purifi...
1733,B01M6453MB,"AFG5FM3NEMOL6BNFRV2NK5FNJCHQ,AGEINTRN6Z563RMLH...","Manu Bhai,Naveenpittu,Evatira Sangma,JAGANNADH...","R3DDL2UPKQ2CK9,R2SYYU1OATVIU5,R1VM993161IYRW,R...","ok,everything was good couldn't return bcoz I ...","ok,got everything as mentioned but the measuri...",2023-12-30 06:05:54.098360656,Prestige Delight PRWO Electric Rice Cooker (1 ...,Home&Kitchen,"₹2,280","₹3,045",4.1,4118.0,"230 Volts, 400 watts, 1 Year",https://m.media-amazon.com/ima

In [ ]:

samples_orders = pd.DataFrame({
    'product_id_1': ['P001', 'P002', 'P003'],
    'user_id': ['U001', 'U002', 'U003'],
    'review_id': ['R001', 'R002', 'R003']
})

samples_products = pd.DataFrame({
    'product_id_2': ['P001', 'P002', 'P004'],
    'product_name': ['Product A', 'Product B', 'Product D'],
    'category': ['Category 1', 'Category 2', 'Category 3']
})

In [ ]:
sample_df= pd.merge(samples_orders, samples_products, left_on='product_id_1', right_on='product_id_2', how='right')

In [ ]:
sample_df

,product_id_1,user_id,review_id,product_id_2,product_name,category
0,P001,U001,R001,P001,Product A,Category 1
1,P002,U002,R002,P002,Product B,Category 2
2,NaN,NaN,NaN,P004,Product D,Category 3


In [ ]:
sample_df.isnull().sum() #how to check for NaNs

,0
product_id_1,1
user_id,1
review_id,1
product_id_2,0
product_name,0
category,0


In [ ]:
sample_df= pd.merge(samples_orders, samples_products[['product_id_2','product_name']], left_on='product_id_1', right_on='product_id_2', how='right')

In [ ]:
sample_df

,product_id_1,user_id,review_id,product_id_2,product_name
0,P001,U001,R001,P001,Product A
1,P002,U002,R002,P002,Product B
2,NaN,NaN,NaN,P004,Product D


In [ ]:
df['review_title'].str.upper()

,review_title
0,"SATISFIED,CHARGING IS REALLY FAST,VALUE FOR MO..."
1,"SATISFIED,CHARGING IS REALLY FAST,VALUE FOR MO..."
2,"SATISFIED,CHARGING IS REALLY FAST,VALUE FOR MO..."
3,"A GOOD BRAIDED CABLE FOR YOUR TYPE C DEVICE,GO..."
4,"A GOOD BRAIDED CABLE FOR YOUR TYPE C DEVICE,GO..."
...,...
1732,"RECEIVED THE PRODUCT WITHOUT SPANNER,EXCELLENT..."
1733,"OK,EVERYTHING WAS GOOD COULDN'T RETURN BCOZ I ..."
1734,"VERY GOOD,WORK BUT FRONT MELT AFTER 2 MONTH,GO..."
1735,"FAN SPEED IS SLOW,GOOD QUALITY,GOOD PRODUCT,GO..."


In [ ]:
df['review_title']

,review_title
0,"Satisfied,Charging is really fast,Value for mo..."
1,"Satisfied,Charging is really fast,Value for mo..."
2,"Satisfied,Charging is really fast,Value for mo..."
3,"A Good Braided Cable for Your Type C Device,Go..."
4,"A Good Braided Cable for Your Type C Device,Go..."
...,...
1732,"Received the product without spanner,Excellent..."
1733,"ok,everything was good couldn't return bcoz I ..."
1734,"very good,Work but front melt after 2 month,Go..."
1735,"Fan Speed is slow,Good quality,Good product,go..."


In [ ]:
def uppercase_title(title):
  #10 operations
    return title.upper()

df['review_title'].apply(uppercase_title)



,review_title
0,"SATISFIED,CHARGING IS REALLY FAST,VALUE FOR MO..."
1,"SATISFIED,CHARGING IS REALLY FAST,VALUE FOR MO..."
2,"SATISFIED,CHARGING IS REALLY FAST,VALUE FOR MO..."
3,"A GOOD BRAIDED CABLE FOR YOUR TYPE C DEVICE,GO..."
4,"A GOOD BRAIDED CABLE FOR YOUR TYPE C DEVICE,GO..."
...,...
1732,"RECEIVED THE PRODUCT WITHOUT SPANNER,EXCELLENT..."
1733,"OK,EVERYTHING WAS GOOD COULDN'T RETURN BCOZ I ..."
1734,"VERY GOOD,WORK BUT FRONT MELT AFTER 2 MONTH,GO..."
1735,"FAN SPEED IS SLOW,GOOD QUALITY,GOOD PRODUCT,GO..."


In [ ]:
x

In [ ]:
def extract_price(price_str):
    try:
        return float(str(price_str).replace('₹', '').replace(',', ''))
    except ValueError:
        return None

In [ ]:
df['actual_price_fixed']= df['actual_price'].apply(extract_price)

In [ ]:
df['discounted_price_fixed']= df['discounted_price'].apply(extract_price)

In [ ]:
df

,product_id,user_id,user_name,review_id,review_title,review_content,order_timestamp,product_name,category,discounted_price,actual_price,rating,rating_count,about_product,img_link,product_link,actual_price_fixed,discounted_price_fixed
0,B07JW9H4J1,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,2023-01-01 00:00:00.000000000,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories,₹399,"₹1,099",4.2,24269.0,High Compatibility : Compatible With iPhone 12...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...,1099.0,399.0
1,B07JW9H4J1,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,2023-01-01 00:00:00.000000000,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories,₹399,"₹1,099",4.2,24270.0,High Compatibility : Compatible With iPhone 12...,https://m.media-amazon.com/images/I/51UsScvHQN...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...,1099.0,399.0
2,B07JW9H4J1,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,2023-01-01 00:00:00.000000000,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories,₹399,"₹1,099",4.2,24269.0,High Compatibility : Compatible With iPhone 12...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...,1099.0,399.0
3,B098NS6PVG,"AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...","ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...","RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...","A Good Braided Cable for Your Type C Device,Go...",I ordered this cable to connect my phone to An...,2023-01-01 05:58:01.967213114,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories,₹199,₹349,4.0,43994.0,"Compatible with all Type C enabled devices, be...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Ambrane-Unbreakable-Char...,349.0,199.0
4,B098NS6PVG,"AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...","ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...","RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...","A Good Braided Cable for Your Type C Device,Go...",I ordered this cable to connect my phone to An...,2023-01-01 05:58:01.967213114,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories,₹199,₹349,4.0,43993.0,"Compatible with all Type C enabled devices, be...",https://m.media-amazon.com/images/I/31zOsqQOAO...,https://www.amazon.in/Ambrane-Unbreakable-Char...,349.0,199.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1732,B08L7J3T31,"AHITFY6AHALOFOHOZEOC6XBP4FEA,AFRABBODZJZQB6Z4U...","Prabha ds,Raghuram bk,Real Deal,Amazon Custome...","R3G3XFHPBFF0E8,R3C0BZCD32EIGW,R2EBVBCN9QPD9R,R...","Received the product without spanner,Excellent...","I received product without spanner,Excellent p...",2023-12-30 00:07:52.131147540,Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...,Home&Kitchen,₹379,₹919,4.0,1090.0,SUPREME QUALITY 90 GRAM 3 LAYER THIK PP SPUN F...,https://m.media-amazon.com/images/I/41fDdRtjfx...,https://www.amazon.in/Noir-Aqua-Spanner-Purifi...,919.0,379.0
1733,B01M6453MB,"AFG5FM3NEMOL6BNFRV2NK5FNJCHQ,AGEINTRN6Z563RMLH...","Manu Bhai,Naveenpittu,Evatira Sangma,JAGANNADH...","R3DDL2UPKQ2CK9,R2SYYU1OATVIU5,R1VM993161IYRW,R...","ok,everything was good couldn't return bcoz I ...","ok,got everything as mentioned but the measuri...",2023-12-30 06:05:54.098360656,Prestige Delight PRWO Electr

In [ ]:
# calculate the discount percentage for the product items
def calculate_discount_percentage(ef):
    return ((ef['actual_price_fixed'] - ef['discounted_price_fixed']) / ef['actual_price_fixed']) * 100

# Apply the function along rows (axis=1)
df['calculated_discount'] = df[['actual_price_fixed','discounted_price_fixed']].apply(calculate_discount_percentage, axis=1)

In [ ]:
df

,product_id,user_id,user_name,review_id,review_title,review_content,order_timestamp,product_name,category,discounted_price,actual_price,rating,rating_count,about_product,img_link,product_link,actual_price_fixed,discounted_price_fixed,calculated_discount
0,B07JW9H4J1,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,2023-01-01 00:00:00.000000000,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories,₹399,"₹1,099",4.2,24269.0,High Compatibility : Compatible With iPhone 12...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...,1099.0,399.0,63.694268
1,B07JW9H4J1,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,2023-01-01 00:00:00.000000000,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories,₹399,"₹1,099",4.2,24270.0,High Compatibility : Compatible With iPhone 12...,https://m.media-amazon.com/images/I/51UsScvHQN...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...,1099.0,399.0,63.694268
2,B07JW9H4J1,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,2023-01-01 00:00:00.000000000,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories,₹399,"₹1,099",4.2,24269.0,High Compatibility : Compatible With iPhone 12...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...,1099.0,399.0,63.694268
3,B098NS6PVG,"AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...","ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...","RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...","A Good Braided Cable for Your Type C Device,Go...",I ordered this cable to connect my phone to An...,2023-01-01 05:58:01.967213114,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories,₹199,₹349,4.0,43994.0,"Compatible with all Type C enabled devices, be...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Ambrane-Unbreakable-Char...,349.0,199.0,42.979943
4,B098NS6PVG,"AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...","ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...","RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...","A Good Braided Cable for Your Type C Device,Go...",I ordered this cable to connect my phone to An...,2023-01-01 05:58:01.967213114,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories,₹199,₹349,4.0,43993.0,"Compatible with all Type C enabled devices, be...",https://m.media-amazon.com/images/I/31zOsqQOAO...,https://www.amazon.in/Ambrane-Unbreakable-Char...,349.0,199.0,42.979943
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1732,B08L7J3T31,"AHITFY6AHALOFOHOZEOC6XBP4FEA,AFRABBODZJZQB6Z4U...","Prabha ds,Raghuram bk,Real Deal,Amazon Custome...","R3G3XFHPBFF0E8,R3C0BZCD32EIGW,R2EBVBCN9QPD9R,R...","Received the product without spanner,Excellent...","I received product without spanner,Excellent p...",2023-12-30 00:07:52.131147540,Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...,Home&Kitchen,₹379,₹919,4.0,1090.0,SUPREME QUALITY 90 GRAM 3 LAYER THIK PP SPUN F...,https://m.media-amazon.com/images/I/41fDdRtjfx...,https://www.amazon.in/Noir-Aqua-Spanner-Purifi...,919.0,379.0,58.759521
1733,B01M6453MB,"AFG5FM3NEMOL6BNFRV2NK5FNJCHQ,AGEINTRN6Z563RMLH...","Manu Bhai,Naveenpittu,Evatira Sangma,JAGANNADH...","R3DDL2UPKQ2CK9,R2SYYU1OATVIU5,R1VM993161IYRW,R...","ok,everything was good couldn't return bcoz I ...","ok,got everything as ment

In [ ]:
import google.generativeai as genai
from google.colab import userdata

genai.configure(api_key=userdata.get('api_key'))
model = genai.GenerativeModel("gemini-2.5-flash")

/usr/local/lib/python3.12/dist-packages/google/colab/_import_hooks/_hook_injector.py:55: FutureWarning: 

All support for the `google.generativeai` package has ended. It will no longer be receiving 
updates or bug fixes. Please switch to the `google.genai` package as soon as possible.
See README for more details:

https://github.com/google-gemini/deprecated-generative-ai-python/blob/main/README.md

  loader.exec_module(module)


In [ ]:
response = model.generate_content(
    contents="""I calculated discount percentage in Pandas using .apply() with axis=1:

df['calculated_discount'] = df.apply(
    lambda row: ((row['actual_price_fixed'] - row['discounted_price_fixed']) / row['actual_price_fixed']) * 100,
    axis=1
)

Show me 3 different ways to achieve the same result in Pandas:
1. The vectorized way (no .apply)
2. Using .eval()
3. Using np.where to handle division by zero

For each, show the code and one line on when you'd prefer it."""
)
print(response.text)

You're right to look for alternatives to `.apply(axis=1)` for arithmetic operations, as it's often much slower than vectorized methods in Pandas.

Here are 3 different ways to achieve the same result, including handling division by zero for one of them:

First, let's set up a sample DataFrame:

```python
import pandas as pd
import numpy as np

# Sample DataFrame
data = {
    'product_id': [1, 2, 3, 4, 5, 6],
    'actual_price_fixed': [100.0, 50.0, 200.0, 75.0, 0.0, 150.0],
    'discounted_price_fixed': [80.0, 50.0, 150.0, 60.0, 0.0, 140.0]
}
df = pd.DataFrame(data)

print("Original DataFrame:")
print(df)
print("-" * 30)
```

### 1. The Vectorized Way (no `.apply()`)

This is the most common and performant way for simple arithmetic operations. Pandas Series natively support vectorized operations, meaning the operation is applied element-wise across entire columns without explicit Python loops.

```python
# 1. The Vectorized Way
df['calculated_discount_vectorized'] = ((df['actual_price_f

In [ ]:
#Business Questions


## How can Amazon analyze average discount percentages by product category to identify which categories
## have the most effective discount strategies for driving sales?

##top 3

df.groupby('category')['calculated_discount'].mean().sort_values(ascending=False).head(3)

,calculated_discount
category,
HomeImprovement,57.944088
Computers&Accessories,55.374270
Health&PersonalCare,52.684211


In [ ]:
#Categories with an average rating of >= 4
cf = df.groupby('category')['rating'].mean()

In [ ]:
cf[cf >= 4]

,rating
category,
Computers&Accessories,4.158934
Electronics,4.090508
Health&PersonalCare,4.000000
Home&Kitchen,4.045890
HomeImprovement,4.250000
OfficeProducts,4.309677
Toys&Games,4.300000


In [ ]:
cf = df.groupby('category')['rating'].mean().reset_index()

In [ ]:
cf

,category,rating
0,Car&Motorbike,3.800000
1,Computers&Accessories,4.158934
2,Electronics,4.090508
3,Health&PersonalCare,4.000000
4,Home&Kitchen,4.045890
5,HomeImprovement,4.250000
6,MusicalInstruments,3.900000
7,OfficeProducts,4.309677
8,Toys&Games,4.300000


In [ ]:
cf[cf['rating']>=4]

,category,rating
1,Computers&Accessories,4.158934
2,Electronics,4.090508
3,Health&PersonalCare,4.000000
4,Home&Kitchen,4.045890
5,HomeImprovement,4.250000
7,OfficeProducts,4.309677
8,Toys&Games,4.300000


In [ ]:

#Questions in notes to try

## Summary of Current Colab Notebook

So far, we have performed the following steps:

1.  **Data Loading**: Downloaded and unzipped `amazon_data4.zip`, then loaded `orders.csv` and `products.csv` into pandas DataFrames.
2.  **Data Merging**: Merged the `orders` and `products` DataFrames into a single `df` DataFrame based on `product_id` using a left merge.
3.  **Data Inspection**: Explored the basic structure, data types, and summary statistics of both `orders` and `products` DataFrames using `.info()` and `.describe()`.
4.  **Data Cleaning & Transformation**:
    *   Defined an `extract_price` function to clean and convert `actual_price` and `discounted_price` columns from string format (e.g., '₹1,099') to numeric (float). New columns `actual_price_fixed` and `discounted_price_fixed` were created.
    *   Calculated the `calculated_discount` percentage for each product based on its `actual_price_fixed` and `discounted_price_fixed`.
5.  **Initial Business Insights**:
    *   Identified the top 3 product categories with the highest average discount percentages.
    *   Identified product categories that have an average rating of 4 or higher.
6.  **LLM Exploration**: Used the Gemini API to explore alternative, more efficient ways to calculate discount percentages in Pandas (vectorized operations, `.eval()`, `np.where`).

## Further Data Insights

### 1. Top 10 Products by Review Count

Let's identify the products that have the most reviews, which can be an indicator of their popularity or market interest.

In [ ]:
# Group by product name and sum the rating counts to find the most reviewed products
top_reviewed_products = df.groupby(['product_name', 'category'])['rating_count'].sum().nlargest(10).reset_index()
display(top_reviewed_products.style.background_gradient(cmap='Blues'))

### 2. Product with the Highest Discount

Understanding which products offer the highest discounts can reveal strategic pricing or clearance items.

In [ ]:
# Find the product with the maximum calculated discount
highest_discount_product = df.loc[df['calculated_discount'].idxmax()]
display(highest_discount_product[['product_name', 'category', 'actual_price_fixed', 'discounted_price_fixed', 'calculated_discount']])